InvalidRequestError: Class does not have a __table__ or __tablename__ specified and does not inherit from an existing table-mapped class.

Hi,

First of all congrats on the new release! This will allow us to go back to a pinned version of flask-sqlalchemy instead of the master branch 👍 

On the other hand, the change #467 now generates this stack-trace in the [Flask-AppBuilder](https://github.com/dpgaspar/Flask-AppBuilder/) project which I would need some help investigating to be able to use release 2.2 directly:
```
12:48:03   File "/src/<module_name>/models.py", line 10, in <module>
12:48:03     from flask_appbuilder.security.sqla.models import User
12:48:03   File "/usr/local/lib/python2.7/dist-packages/flask_appbuilder/security/sqla/models.py", line 12, in <module>
12:48:03     class Permission(Model):
12:48:03   File "/usr/local/lib/python2.7/dist-packages/flask_sqlalchemy/__init__.py", line 602, in __init__
12:48:03     DeclarativeMeta.__init__(self, name, bases, d)
12:48:03   File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/ext/declarative/api.py", line 55, in __init__
12:48:03     _as_declarative(cls, classname, cls.__dict__)
12:48:03   File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/ext/declarative/base.py", line 88, in _as_declarative
12:48:03     _MapperConfig.setup_mapping(cls, classname, dict_)
12:48:03   File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/ext/declarative/base.py", line 103, in setup_mapping
12:48:03     cfg_cls(cls_, classname, dict_)
12:48:03   File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/ext/declarative/base.py", line 133, in __init__
12:48:03     self._setup_inheritance()
12:48:03   File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/ext/declarative/base.py", line 429, in _setup_inheritance
12:48:03     "table-mapped class." % cls
12:48:03 InvalidRequestError: Class <class 'flask_appbuilder.security.sqla.models.Permission'> does not have a __table__ or __tablename__ specified and does not inherit from an existing table-mapped class.
```

The error message contradicts the [model declaration](https://github.com/dpgaspar/Flask-AppBuilder/blob/master/flask_appbuilder/security/sqla/models.py#L12-L18) as it does have a `__tablename` defined:
```python
class Permission(Model):
    __tablename__ = 'ab_permission'
    id = Column(Integer, Sequence('ab_permission_id_seq'), primary_key=True)
    name = Column(String(100), unique=True, nullable=False)

    def __repr__(self):
        return self.name
```

I understand that this change is related to the `@declared_attr` decorator but despite reading http://docs.sqlalchemy.org/en/latest/orm/extensions/declarative/api.html#api-reference I don't really know exactly how I should apply it to the model.

Thanks for your help

@davidism This does seem to be an issue. Using a __tablename__ declared_attr from a 'mixin' does not work:

```
    def test_tablename_mixin(self):
        app = flask.Flask(__name__)
        db = fsa.SQLAlchemy(app)

        class Duck(db.Model):
            __abstract__ = True

        class NameMixin():

            @declared_attr
            def __tablename__(cls):
                return 'drake_mallard'

        class Darkwing(Duck, NameMixin):
            id = db.Column(db.Integer, primary_key=True)

        class Donald(Duck):
            id = db.Column(db.Integer, primary_key=True)

        self.assertEqual(Darkwing.__tablename__, 'drake_mallard')
        self.assertEqual(Donald.__tablename__, 'donald')
```

Results in the following test failure:
```
======================================================================
FAIL: test_tablename_mixin (__main__.TablenameTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_sqlalchemy.py", line 359, in test_tablename_mixin
    self.assertEqual(Darkwing.__tablename__, 'drake_mallard')
AssertionError: 'darkwing' != 'drake_mallard'
- darkwing
+ drake_mallard
```

I do notice in the execution the following error is spawned:
```
sqlalchemy/ext/declarative/api.py:182: SAWarning: Unmanaged access of declarative attribute __tablename__ from non-mapped class Base
  (desc.fget.__name__, cls.__name__))
```

Perhaps I'm using my name mixin wrong?

--- a/flask_sqlalchemy/__init__.py
+++ b/flask_sqlalchemy/__init__.py
@@ -30,7 +30,7 @@ from sqlalchemy.ext.declarative import DeclarativeMeta, declarative_base, \
 from sqlalchemy.orm.exc import UnmappedClassError
 from sqlalchemy.orm.session import Session as SessionBase
 
-from ._compat import iteritems, itervalues, string_types, xrange
+from ._compat import itervalues, string_types, xrange
 
 __version__ = '2.2.1'
 
@@ -551,31 +551,39 @@ class _EngineConnector(object):
 
 
 def _should_set_tablename(cls):
-    """Traverse the model's MRO. If a primary key column is found before a
-    table or tablename, then a new tablename should be generated.
+    """Determine whether ``__tablename__`` should be automatically generated
+    for a model.
 
-    This supports:
+    * If no class in the MRO sets a name, one should be generated.
+    * If a declared attr is found, it should be used instead.
+    * If a name is found, it should be used if the class is a mixin, otherwise
+      one should be generated.
+    * Abstract models should not have one generated.
 
-    * Joined table inheritance without explicitly naming sub-models.
-    * Single table inheritance.
-    * Inheriting from mixins or abstract models.
-
-    :param cls: model to check
-    :return: True if tablename should be set
+    Later, :meth:`._BoundDeclarativeMeta.__table_cls__` will determine if the
+    model looks like single or joined-table inheritance. If no primary key is
+    found, the name will be unset.
     """
+    if (
+        cls.__dict__.get('__abstract__', False)
+        or not any(isinstance(b, DeclarativeMeta) for b in cls.__mro__[1:])
+    ):
+        return False
 
     for base in cls.__mro__:
-        d = base.__dict__
+        if '__tablename__' not in base.__dict__:
+            continue
 
-        if '__tablename__' in d or '__table__' in d:
+        if isinstance(base.__dict__['__tablename__'], declared_attr):
             return False
 
-        for name, obj in iteritems(d):
-            if isinstance(obj, declared_attr):
-                obj = getattr(cls, name)
+        return not (
+            base is cls
+            or base.__dict__.get('__abstract__', False)
+            or not isinstance(base, DeclarativeMeta)
+        )
 
-            if isinstance(obj, sqlalchemy.Column) and obj.primary_key:
-                return True
+    return True
 
 
 def camel_to_snake_case(name):
@@ -591,20 +599,36 @@ def camel_to_snake_case(name):
 
 
 class _BoundDeclarativeMeta(DeclarativeMeta):
-    def __new__(cls, name, bases, d):
-        # if tablename is set explicitly, move it to the cache attribute so
-        # that future subclasses still have auto behavior
-        if '__tablename__' in d:
-            d['_cached_tablename'] = d.pop('__tablename__')
+    def __init__(cls, name, bases, d):
+        if _should_set_tablename(cls):
+            cls.__tablename__ = camel_to_snake_case(cls.__name__)
 
-        return DeclarativeMeta.__new__(cls, name, bases, d)
+        bind_key = (
+            d.pop('__bind_key__', None)
+            or getattr(cls, '__bind_key__', None)
+        )
 
-    def __init__(self, name, bases, d):
-        bind_key = d.pop('__bind_key__', None) or getattr(self, '__bind_key__', None)
-        DeclarativeMeta.__init__(self, name, bases, d)
+        super(_BoundDeclarativeMeta, cls).__init__(name, bases, d)
 
-        if bind_key is not None and hasattr(self, '__table__'):
-            self.__table__.info['bind_key'] = bind_key
+        if bind_key is not None and hasattr(cls, '__table__'):
+            cls.__table__.info['bind_key'] = bind_key
+
+    def __table_cls__(cls, *args, **kwargs):
+        """This is called by SQLAlchemy during mapper setup. It determines the
+        final table object that the model will use.
+
+        If no primary key is found, that indicates single-table inheritance,
+        so no table will be created and ``__tablename__`` will be unset.
+        """
+        for arg in args:
+            if (
+                (isinstance(arg, sqlalchemy.Column) and arg.primary_key)
+                or isinstance(arg, sqlalchemy.PrimaryKeyConstraint)
+            ):
+                return sqlalchemy.Table(*args, **kwargs)
+
+        if '__tablename__' in cls.__dict__:
+            del cls.__tablename__
 
 
 def get_state(app):
@@ -637,18 +661,6 @@ class Model(object):
     #: Convenience property to query the database for instances of this model using the current session.
     #: Equivalent to ``db.session.query(Model)`` unless :attr:`query_class` has been changed.
     query = None
-
-    _cached_tablename = None
-
-    @declared_attr
-    def __tablename__(cls):
-        if (
-            '_cached_tablename' not in cls.__dict__ and
-            _should_set_tablename(cls)
-        ):
-            cls._cached_tablename = camel_to_snake_case(cls.__name__)
-
-        return cls._cached_tablename
 
 
 class SQLAlchemy(object):

--- a/tests/test_table_name.py
+++ b/tests/test_table_name.py
@@ -1,3 +1,5 @@
+import inspect
+
 from sqlalchemy.ext.declarative import declared_attr
 
 
@@ -25,6 +27,7 @@ def test_single_name(db):
     class Mallard(Duck):
         pass
 
+    assert '__tablename__' not in Mallard.__dict__
     assert Mallard.__tablename__ == 'duck'
 
 
@@ -39,8 +42,10 @@ def test_joined_name(db):
     assert Donald.__tablename__ == 'donald'
 
 
-def test_mixin_name(db):
-    """Primary key provided by mixin should still allow model to set tablename."""
+def test_mixin_id(db):
+    """Primary key provided by mixin should still allow model to set
+    tablename.
+    """
     class Base(object):
         id = db.Column(db.Integer, primary_key=True)
 
@@ -51,8 +56,33 @@ def test_mixin_name(db):
     assert Duck.__tablename__ == 'duck'
 
 
+def test_mixin_attr(db):
+    """A declared attr tablename will be used down multiple levels of
+    inheritance.
+    """
+    class Mixin(object):
+        @declared_attr
+        def __tablename__(cls):
+            return cls.__name__.upper()
+
+    class Bird(Mixin, db.Model):
+        id = db.Column(db.Integer, primary_key=True)
+
+    class Duck(Bird):
+        # object reference
+        id = db.Column(db.ForeignKey(Bird.id), primary_key=True)
+
+    class Mallard(Duck):
+        # string reference
+        id = db.Column(db.ForeignKey('DUCK.id'), primary_key=True)
+
+    assert Bird.__tablename__ == 'BIRD'
+    assert Duck.__tablename__ == 'DUCK'
+    assert Mallard.__tablename__ == 'MALLARD'
+
+
 def test_abstract_name(db):
-    """Abstract model should not set a name.  Subclass should set a name."""
+    """Abstract model should not set a name. Subclass should set a name."""
     class Base(db.Model):
         __abstract__ = True
         id = db.Column(db.Integer, primary_key=True)
@@ -60,19 +90,23 @@ def test_abstract_name(db):
     class Duck(Base):
         pass
 
-    assert Base.__tablename__ == 'base'
+    assert '__tablename__' not in Base.__dict__
     assert Duck.__tablename__ == 'duck'
 
 
 def test_complex_inheritance(db):
-    """Joined table inheritance, but the new primary key is provided by a mixin, not directly on the class."""
+    """Joined table inheritance, but the new primary key is provided by a
+    mixin, not directly on the class.
+    """
     class Duck(db.Model):
         id = db.Column(db.Integer, primary_key=True)
 
     class IdMixin(object):
         @declared_attr
         def id(cls):
-            return db.Column(db.Integer, db.ForeignKey(Duck.id), primary_key=True)
+            return db.Column(
+                db.Integer, db.ForeignKey(Duck.id), primary_key=True
+            )
 
     class RubberDuck(IdMixin, Duck):
         pass
@@ -81,18 +115,55 @@ def test_complex_inheritance(db):
 
 
 def test_manual_name(db):
+    """Setting a manual name prevents generation for the immediate model. A
+    name is generated for joined but not single-table inheritance.
+    """
     class Duck(db.Model):
         __tablename__ = 'DUCK'
         id = db.Column(db.Integer, primary_key=True)
+        type = db.Column(db.String)
+
+        __mapper_args__ = {
+            'polymorphic_on': type
+        }
 
     class Daffy(Duck):
         id = db.Column(db.Integer, db.ForeignKey(Duck.id), primary_key=True)
 
+        __mapper_args__ = {
+            'polymorphic_identity': 'Warner'
+        }
+
+    class Donald(Duck):
+        __mapper_args__ = {
+            'polymorphic_identity': 'Disney'
+        }
+
     assert Duck.__tablename__ == 'DUCK'
     assert Daffy.__tablename__ == 'daffy'
+    assert '__tablename__' not in Donald.__dict__
+    assert Donald.__tablename__ == 'DUCK'
+    # polymorphic condition for single-table query
+    assert 'WHERE "DUCK".type' in str(Donald.query)
+
+
+def test_primary_constraint(db):
+    """Primary key will be picked up from table args."""
+    class Duck(db.Model):
+        id = db.Column(db.Integer)
+
+        __table_args__ = (
+            db.PrimaryKeyConstraint(id),
+        )
+
+    assert Duck.__table__ is not None
+    assert Duck.__tablename__ == 'duck'
 
 
 def test_no_access_to_class_property(db):
+    """Ensure the implementation doesn't access class properties or declared
+    attrs while inspecting the unmapped model.
+    """
     class class_property(object):
         def __init__(self, f):
             self.f = f
@@ -106,14 +177,13 @@ def test_no_access_to_class_property(db):
     class ns(object):
         accessed = False
 
-    # Since there's no id provided by the following model,
-    # _should_set_tablename will scan all attributes. If it's working
-    # properly, it won't access the class property, but will access the
-    # declared_attr.
-
     class Witch(Duck):
         @declared_attr
         def is_duck(self):
+            # declared attrs will be accessed during mapper configuration,
+            # but make sure they're not accessed before that
+            info = inspect.getouterframes(inspect.currentframe())[2]
+            assert info[3] != '_should_set_tablename'
             ns.accessed = True
 
         @class_property

